torch/cuda/streams.py

```diff
@@ -119,90 +102,75 @@ def __repr__(self):
 
-    DEFAULT = 0x0
-    BLOCKING_SYNC = 0x1
-    DISABLE_TIMING = 0x2
-    INTERPROCESS = 0x4
-
-    def __init__(self, enable_timing=False, blocking=False, interprocess=False,
-                 _handle=None):
-        flags = Event.DEFAULT
-        if not enable_timing:
-            flags |= Event.DISABLE_TIMING
-        if blocking:
-            flags |= Event.BLOCKING_SYNC
-        if interprocess:
-            flags |= Event.INTERPROCESS
-
-        ptr = ctypes.c_void_p()
-        self._cudart = cudart()
-        if _handle:
-            check_error(self._cudart.cudaIpcOpenEventHandle(ctypes.byref(ptr), _handle))
-        else:
-            check_error(self._cudart.cudaEventCreateWithFlags(ctypes.byref(ptr), ctypes.c_uint(flags)))
-        self._as_parameter_ = ptr
-
-    def __del__(self):
-        if hasattr(self, '_as_parameter_'):
-            check_error(self._cudart.cudaEventDestroy(self._as_parameter_))
-            del self._as_parameter_
+    def __new__(cls, enable_timing=False, blocking=False, interprocess=False,
+                **kwargs):
+        return super(Event, cls).__new__(
+            cls, enable_timing=enable_timing, blocking=blocking,
+            interprocess=interprocess, **kwargs)
 
     def record(self, stream=None):
-        r"""Records the event in a given stream."""
+        r"""Records the event in a given stream. Use
+        ``torch.cuda.current_stream()`` if not given. The stream's device must agree
+        with event's device."""
         if stream is None:
             stream = torch.cuda.current_stream()
-        stream.record_event(self)
+        super(Event, self).record(stream)
 
     def wait(self, stream=None):
-        r"""Makes a given stream wait for the event."""
+        r"""Makes a given stream wait for the event. Use
+        ``torch.cuda.current_stream()`` is not given."""
         if stream is None:
             stream = torch.cuda.current_stream()
-        stream.wait_event(self)
+        super(Event, self).wait(stream)
 
     def query(self):
-        r"""Checks if the event has been recorded.
+        r"""Checks if all work currently captured by event has completed. This
+        can be called on any device.
 
         Returns:
-            A boolean indicating if the event has been recorded.
+            A boolean indicating if all work currently captured by event has
+            completed.
         """
-        res = cudart().cudaEventQuery(self)
-        if res == cudaStatus.ERROR_NOT_READY:
-            return False
-        check_error(res)
-        return True
+        return super(Event, self).query()
 
     def elapsed_time(self, end_event):
-        r"""Returns the time elapsed in milliseconds before the event was recorded."""
-        time_ms = ctypes.c_float()
-        check_error(cudart().cudaEventElapsedTime(
-            ctypes.byref(time_ms), self, end_event))
-        return time_ms.value
+        r"""Returns the time elapsed in milliseconds after the event was
```

Comment:
The events can be recorded on different devices. Suggestion:

```python
r"""Returns the elapsed time in milliseconds between two events.

Both events must be created with `enable_timing=True` and have already be recorded and completed (that is `event.query()` must be True).

NOTE: If either event was last recorded in a non-NULL stream, the resulting time may be greater than expected (even if both used the same stream).

        .. note:: This is a wrapper around ``cudaEventElapsedTime()``: see `CUDA
           documentation`_ for more info.

        .. _CUDA documentation:
           https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html
"""
```

Reply:
I thought I tests two events on different devices, and it hit CUDA error. Let me double check.

Reply:
The following code run into `RuntimeError: CUDA error: invalid resource handle` error. But I can still remove that from doc as CUDA does not explicitly say that.

```python
        d0 = torch.device('cuda:0')
        d1 = torch.device('cuda:1')
    
        with torch.cuda.device(d0):
            s0 = torch.cuda.current_stream()
            e0 = torch.cuda.Event(enable_timing=True)
            torch.cuda._sleep(10)  # spin for about 50 ms on device1
            s0.record_event(e0)
    
        with torch.cuda.device(d1):
            s1 = torch.cuda.current_stream()
            e1 = torch.cuda.Event(enable_timing=True)
            torch.cuda._sleep(50000000)  # spin for about 50 ms on device1
            s1.record_event(e1)
    
        e0.synchronize()
        e1.synchronize()
       e0.elapsed_time(e1)
```